flink-core/src/main/java/org/apache/flink/api/common/TaskInfo.java

```diff
@@ -67,11 +56,12 @@ public TaskInfo(
 		checkArgument(indexOfSubtask < numberOfParallelSubtasks, "Task index must be less than parallelism.");
 		checkArgument(attemptNumber >= 0, "Attempt number must be a non-negative number.");
 		this.taskName = checkNotNull(taskName, "Task Name must not be null.");
+		this.taskNameWithSubtasks = checkNotNull(taskNameWithSubtasks);
+		this.taskNameWithSubtaskAndID = checkNotNull(taskNameWithSubtaskAndID);
```

Comment:
This seems a bit odd. We are storing three times the task name. I think this is duplicated information. I think it would be better to rather store the `taskName`, its index, the number of maximum tasks and its `ExecutionAttemptID`. If `TaskInfo` contains this, then we could generate the `taskNameWithSubtaskAndID` based on this information when instantiating the `One/TwoInputStreamTask`.

Reply:
Yes, that is another option, but that would cause generate this task name multiple times for `BarrierBuffer` and `StreamProcessor`. The benefit of generating task name in `TaskInfo` is that it is  done only once and be reused for all the places. If we further extend to reuse this name in other components we also do not need to generate it again.

Reply:
Ok, then I would not modify the structure of `TaskInfo` and only instantiate the task name in separate components via `RuntimeEnvironment#getExecutionID`.